core/src/main/scala/kafka/controller/ControllerRequestMerger.scala

```diff
@@ -0,0 +1,245 @@
+  var controllerState : RequestControllerState = null
+
+  // Here we store one callback for the LeaderAndIsr response and one for the StopReplica response,
+  // given all the requests of a particular type sent to the same broker always have the same callback.
+  var leaderAndIsrCallback: AbstractResponse => Unit = null
+  var stopReplicaCallback: AbstractResponse => Unit = null
+
+  def addRequest(request: AbstractControlRequest.Builder[_ <: AbstractControlRequest],
+    callback: AbstractResponse => Unit = null): Unit = {
+    val newControllerState = new RequestControllerState(request.controllerId(), request.controllerEpoch())
+    if (controllerState != null) {
+      if (!controllerState.equals(newControllerState)) {
+        throw new IllegalStateException("The controller state in the ControllerRequestMerger should not change")
+      }
+    } else {
+      controllerState = newControllerState
+    }
+
+    request match {
+      case leaderAndIsrRequest : LeaderAndIsrRequest.Builder => addLeaderAndIsrRequest(leaderAndIsrRequest, callback)
+      case updateMetadataRequest : UpdateMetadataRequest.Builder => addUpdateMetadataRequest(updateMetadataRequest)
+      case stopReplicaRequest: StopReplicaRequest.Builder => addStopReplicaRequest(stopReplicaRequest, callback)
+    }
+  }
+
+  def isLeaderAndIsrReplaceable(newState: LeaderAndIsrPartitionState, currentState: LeaderAndIsrPartitionState): Boolean = {
+    newState.brokerEpoch() > currentState.brokerEpoch() || newState.leaderEpoch() > currentState.leaderEpoch()
+  }
+
+  def mergeLeaderAndIsrPartitionState(incomingState: LeaderAndIsrPartitionState,
+    queuedStates: util.LinkedList[LeaderAndIsrPartitionState]): Unit = {
+    // keep merging requests from the tail of the queued States
+    while (!queuedStates.isEmpty && isLeaderAndIsrReplaceable(incomingState, queuedStates.getLast)) {
+      queuedStates.pollLast()
+    }
+    val inserted = queuedStates.offerLast(incomingState)
+    if (!inserted) {
+      throw new IllegalStateException(s"Unable to insert LeaderAndIsrPartitionState $incomingState to the merger queue")
+    }
+  }
+
+  private def addLeaderAndIsrRequest(request: LeaderAndIsrRequest.Builder,
+    callback: AbstractResponse => Unit): Unit = {
+    request.partitionStates().forEach{partitionState => {
+      val transformedPartitionState = LiCombinedControlTransformer.transformLeaderAndIsrPartition(partitionState, request.maxBrokerEpoch())
+
+      val topicPartition = new TopicPartition(partitionState.topicName(), partitionState.partitionIndex())
+      val queuedStates = leaderAndIsrPartitionStates.getOrElseUpdate(topicPartition,
+        new util.LinkedList[LeaderAndIsrPartitionState]())
+
+      mergeLeaderAndIsrPartitionState(transformedPartitionState, queuedStates)
+
+      // one LeaderAndIsr request renders the previous StopReplica requests non-applicable
+      clearStopReplicaPartitionState(topicPartition)
```

Comment:
what happens if there is a topic deletion followed by a new topic creation, 
in practice, this should not happen, right? new topic creation won't happen until previous topic deletion ops are completed. 
Is there a race between PLE and topic deletion,  partition reassignment and topic deletion? my understanding is topic deletion itself is protected against other operations.  

Stop replica has two states: delete replica and stop only. 
I am more concerned about "delete replicas" (topic deletion or move replica out by partition reassignment) requests been removed here,  but maybe it is ok as long as it is not topic deletion?

Reply:
1. Yes, deletion of a topic must have been finished completely before the client should create the topic again.
2. Topic deletion is mutually exclusive of PLE or reassignment, so if a topic is already queued for deletion, the PLE and reassignment requests will be ignored.
3. While replying to your comment, I realize there might be a risk for the LeaderAndIsr and StopReplica requests to cancel each other. One hypothetical example is a StopReplica request followed by a LeaderAndIsr (become leader). In that case, if the StopReplica is canceled, maybe the broker becomes the leader while still fetching from the previous leader. To avoid such cases, I'll remove the logic for cancellation between LeaderAndIsr and StopReplica requests.